modules/csg/csg_shape.cpp

```diff
@@ -412,38 +416,25 @@ void CSGShape3D::_update_shape() {
 	//create surfaces
 
 	for (int i = 0; i < surfaces.size(); i++) {
-		// calculate tangents for this surface
-		bool have_tangents = calculate_tangents;
-		if (have_tangents) {
-			SMikkTSpaceInterface mkif;
-			mkif.m_getNormal = mikktGetNormal;
-			mkif.m_getNumFaces = mikktGetNumFaces;
-			mkif.m_getNumVerticesOfFace = mikktGetNumVerticesOfFace;
-			mkif.m_getPosition = mikktGetPosition;
-			mkif.m_getTexCoord = mikktGetTexCoord;
-			mkif.m_setTSpace = mikktSetTSpaceDefault;
-			mkif.m_setTSpaceBasic = nullptr;
-
-			SMikkTSpaceContext msc;
-			msc.m_pInterface = &mkif;
-			msc.m_pUserData = &surfaces.write[i];
-			have_tangents = genTangSpaceDefault(&msc);
-		}
-
 		if (surfaces[i].last_added == 0) {
 			continue;
 		}
+		Ref<SurfaceTool> surface_tool;
 
-		// and convert to surface array
-		Array array;
-		array.resize(Mesh::ARRAY_MAX);
+		surface_tool.instantiate();
+		surface_tool->begin(Mesh::PRIMITIVE_TRIANGLES);
 
-		array[Mesh::ARRAY_VERTEX] = surfaces[i].vertices;
-		array[Mesh::ARRAY_NORMAL] = surfaces[i].normals;
-		array[Mesh::ARRAY_TEX_UV] = surfaces[i].uvs;
-		if (have_tangents) {
-			array[Mesh::ARRAY_TANGENT] = surfaces[i].tans;
+		for (int j = 0; j < surfaces[i].last_added; j++) {
+			surface_tool->set_normal(surfaces[i].normalsw[j]);
+			surface_tool->set_uv(surfaces[i].uvsw[j]);
+			surface_tool->add_vertex(surfaces[i].verticesw[j]);
 		}
+		for (int index_i = 0; index_i < surfaces[i].indices.size(); index_i++) {
+			surface_tool->add_index(surfaces[i].indices[index_i]);
```

Comment:
Above you use the pointer, but the vector here, should it be used here as well?

Reply:
The structure ShapeUpdateSurface is not surface tool? What do you mean?

Reply:
I mean that above you use the ptrw, like so:
```cpp
surface_tool->add_vertex(surfaces[i].verticesw[j]);
```
But here the vector, should it be:
```suggestion
			surface_tool->add_index(surfaces[i].indicesw[index_i]);
```

Reply:
Removed vertex indices of this.